examples/source_separation/utils/metrics.py

```diff
@@ -0,0 +1,164 @@
+        )
+        for i, idx in enumerate(permutations(range(num_speakers))):
+            util = self.utility_func(estimate, reference[:, idx, :])
+            util_mat[:, i] = util.mean(dim=1)  # take the average over speaker dimension
+        return util_mat.max(dim=1).values
+
+
+_sdr_pit = PIT(utility_func=sdr)
+
+
+def sdr_pit(estimate, reference):
+    """Computes scale-invariant source-to-distortion ratio.
+
+    1. adjust both estimate and reference to have 0-mean
+    2. scale the reference signal with power(s_est * s_ref) / powr(s_ref * s_ref)
+    3. compute SNR between adjusted estimate and reference.
+
+    Args:
+        estimate (torch.Tensor): Estimtaed signal.
+            Shape: [batch, speakers (can be 1), time frame]
+        reference (torch.Tensor): Reference signal.
+            Shape: [batch, speakers, time frame]
+        epsilon (float): constant value used to stabilize division.
+
+    Returns:
+        torch.Tensor: scale-invariant source-to-distortion ratio.
+        Shape: [batch, speaker]
+
+    References:
+        - Single-channel multi-speaker separation using deep clustering
+          Y. Isik, J. Le Roux, Z. Chen, S. Watanabe, and J. R. Hershey,
+        - Conv-TasNet: Surpassing Ideal Time--Frequency Magnitude Masking for Speech Separation
+          Luo, Yi and Mesgarani, Nima
+          https://arxiv.org/abs/1809.07454
+
+    Notes:
+        This function is tested to produce the exact same result as the reference implementation,
+        *when the inputs have 0-mean*
+        https://github.com/naplab/Conv-TasNet/blob/e66d82a8f956a69749ec8a4ae382217faa097c5c/utility/sdr.py#L107-L153
+    """
+    return _sdr_pit(estimate, reference)
+
+
+def sdri(estimate: torch.Tensor, reference: torch.Tensor, mix: torch.Tensor) -> torch.Tensor:
+    """Compute the improvement of SDR (SDRi).
+
+    This function compute how much SDR is improved if the estimation is changed from
+    the original mixture signal to the actual estimated source signals. That is,
+    ``SDR(estimate, reference) - SDR(mix, reference)``.
+
+    For computing ``SDR(estimate, reference)``, PIT (permutation invariant training) is applied,
+    so that best combination of sources between the reference signals and the esimate signals
+    are picked.
+
+    Args:
+        estimate (torch.Tensor): Estimated source signals.
+            Shape: [batch, speakers, time frame]
+        reference (torch.Tensor): Reference (original) source signals.
+            Shape: [batch, speakers, time frame]
+        mix (torch.Tensor): Mixed souce signals, from which the setimated signals were generated.
+            Shape: [batch, speakers == 1, time frame]
+
+    Returns:
+        torch.Tensor: Improved SDR. Shape: [batch, ]
+
+    References:
+        - Conv-TasNet: Surpassing Ideal Time--Frequency Magnitude Masking for Speech Separation
+          Luo, Yi and Mesgarani, Nima
+          https://arxiv.org/abs/1809.07454
```

Comment:
I think the first reference should be https://arxiv.org/abs/1811.02508

Reply:
Thanks for the suggestion.

Reply:
@faroit 

I was looking at it, and just realized that SDR is very common metric that may be not novel here.
What do you think I should site? In this PR, I have the following functions;

1. `SDR` which implementation was taken from [here](https://github.com/naplab/Conv-TasNet/blob/e66d82a8f956a69749ec8a4ae382217faa097c5c/utility/sdr.py#L34-L56), which is a part of reference implementation of conv-tasnet provided by the paper authors.
2. `PIT`: reference is the paper https://arxiv.org/abs/1703.06284, which I think is correct.
3. SDR-PIT: I sited the conv-tasnet paper, but is there a better source?
4. `sdri`, which contains the improvement `sdr`, which as you suggested is refered in https://arxiv.org/abs/1811.02508.

Reply:
I think things are a bit messed up with respect to SDR namings (also pointed out in https://arxiv.org/abs/1811.02508)

1. The SDR as implemented [here](https://github.com/naplab/Conv-TasNet/blob/e66d82a8f956a69749ec8a4ae382217faa097c5c/utility/sdr.py#L34-L56) is scale-invariant, so this would should be called `SI-SDR` citing https://arxiv.org/abs/1811.02508
2. agreed
3. I am not sure if there was another paper before, so probably correct. @mpariente ?
4. yes

I would recommend to separate the actual loss function (SDR, SISDR, SDSDR) from the property of being permutation invariant or not (PIT vs non-pit). E.g. like in asteroids [pit wrapper](https://github.com/mpariente/asteroid/blob/master/asteroid/losses/pit_wrapper.py)

Reply:
@faroit Thanks for the reply. Fortunately these metrics went into `examples` directly. So we do can fix this without being worried about backward compatibility. I will open another PR to fix this and ping you.
We are working on the next release (expected to happen at the end of October), so I will do it after that.